libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
 
     self.flags.incl(ServerFlags.ReusePort)
-    let server = createStreamServer(
-      ma = ma,
-      flags = self.flags,
-      udata = self)
 
-    # always get the resolved address in case we're bound to 0.0.0.0:0
-    self.addrs[i] = MultiAddress.init(
-      server.sock.getLocalAddress()
-    ).tryGet()
+    var supported: seq[MultiAddress]
+    var initialized = false
+    try:
+      for i, ma in addrs:
+        if not self.handles(ma):
+          trace "Invalid address detected, skipping!", address = ma
+          continue
 
-    self.servers &= server
+        let
+          ta = initTAddress(ma).expect("valid address per handles check above")
+          server =
+            try:
+              createStreamServer(ta, flags = self.flags)
+            except common.TransportError as exc:
+              raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+
+        self.servers &= server
+
+        trace "Listening on", address = ma
+        supported.add(
+          MultiAddress.init(server.sock.getLocalAddress()).expect(
+            "Can init from local address"
+          )
+        )
+
+      initialized = true
+    finally:
+      if not initialized:
+        # Clean up partial success on exception
+        await noCancel allFutures(self.servers.mapIt(it.closeWait()))
+        reset(self.servers)
 
-    trace "Listening on", address = ma
+    try:
+      await procCall Transport(self).start(supported)
+    except CatchableError:
+      raiseAssert "Base method does not raise"
 
-method stop*(self: TcpTransport) {.async.} =
-  ## stop the transport
-  ##
-  try:
-    trace "Stopping TCP transport"
+    trackCounter(TcpTransportTrackerName)
 
-    checkFutures(
-      await allFinished(
-        self.clients[Direction.In].mapIt(it.closeWait()) &
-        self.clients[Direction.Out].mapIt(it.closeWait())))
+  impl(self, addrs)
 
-    if not self.running:
+method stop*(self: TcpTransport): Future[void] =
+  ## Stop the transport and close all connections it created
+  proc impl(self: TcpTransport) {.async: (raises: []).} =
+    trace "Stopping TCP transport"
+    self.stopping = true
+    defer:
+      self.stopping = false
+
+    if self.running:
+      # Reset the running flag
+      try:
+        await noCancel procCall Transport(self).stop()
+      except CatchableError: # TODO remove when `accept` is annotated with raises
+        raiseAssert "doesn't actually raise"
+
+      # Stop each server by closing the socket - this will cause all accept loops
+      # to fail - since the running flag has been reset, it's also safe to close
+      # all known clients since no more of them will be added
+      await noCancel allFutures(
+        self.servers.mapIt(it.closeWait()) &
+          self.clients[Direction.In].mapIt(it.closeWait()) &
+          self.clients[Direction.Out].mapIt(it.closeWait())
+      )
+
+      self.servers = @[]
+
+      for acceptFut in self.acceptFuts:
+        if acceptFut.completed():
```

Comment:
Don't we want to cancel it when it's not completed?

Reply:
no - only the owner of a future should cancel it - in this case, `accept` is the owner of the futures in `acceptFut` so only accept should cancel them.. 

this is something of  a grey zone in the code, ie the act of closing the transport will cause all accept futures to fail _except_ those that have already been completed, so `accept` does not strictly need to clean up `acceptFuts` even if it "owns" them.

cleaning this up would require changing the api which would require much more extensive changes, thus better left for another PR